bool all
bool left
bool right